src/vm/threads.cpp

```diff
@@ -6009,6 +6042,80 @@ void ThreadStore::TransferStartedThread(Thread *thread, BOOL bRequiresTSL)
+{
+    // Although all increments and decrements are usually done inside a lock, that is not sufficient to synchronize with a
+    // background GC thread resetting this value, hence the interlocked operation. Ignore overflow; overflow would likely never
+    // happen, the count is treated as unsigned, and nothing bad would happen if it were to overflow.
+    SIZE_T count = static_cast<SIZE_T>(FastInterlockIncrement(&m_DeadThreadCountForGCTrigger));
+
+    SIZE_T countThreshold = static_cast<SIZE_T>(s_DeadThreadCountThresholdForGCTrigger);
+    if (count < countThreshold || countThreshold == 0)
+    {
+        return;
+    }
+
+    IGCHeap *gcHeap = GCHeapUtilities::GetGCHeap();
+    SIZE_T millisecondsSinceLastFinalizingGC =
+        gcHeap->GetNow() - gcHeap->GetLastGCStartTime(gcHeap->GetMinFinalizingGeneration());
+    if (millisecondsSinceLastFinalizingGC < s_DeadThreadGCTriggerPeriodMilliseconds)
+    {
+        return;
+    }
+
+    if (!g_fEEStarted)
+    {
+        return;
+    }
+
+    // The dead thread count used for triggering a GC is tracked separately from the actual dead thread count so that a dead
+    // thread does not contribute to triggering a GC more than once. The count exceeds a certain threshold and a finalizing GC
+    // has not occurred in a certain duration, so reset the count and trigger a GC now to finalize unreachable dead threads.
+    // The GC is triggered on the finalizer thread since it's not safe to trigger it on DLL_THREAD_DETACH.
+    m_DeadThreadCountForGCTrigger = 0;
+    m_TriggerFinalizingGC = true;
+    FinalizerThread::EnableFinalization();
+}
+
+void ThreadStore::DecrementDeadThreadCountForGCTrigger()
+{
+    // Although all increments and decrements are usually done inside a lock, that is not sufficient to synchronize with a
+    // background GC thread resetting this value, hence the interlocked operation.
+    if (FastInterlockDecrement(&m_DeadThreadCountForGCTrigger) < 0)
+    {
+        FastInterlockExchange(&m_DeadThreadCountForGCTrigger, 0);
+    }
+}
+
+void ThreadStore::OnFinalizingGCStarted()
+{
+    // A dead thread may contribute to triggering at most once. After a finalizing GC occurs, if some dead thread objects are
+    // still reachable due to references to the thread objects, they will not contribute to triggering a GC again. Synchronize
+    // the store with increment/decrement operations occurring on different threads, and make the change visible to other
+    // threads in order to prevent unnecessary GC triggers.
+    FastInterlockExchange(&m_DeadThreadCountForGCTrigger, 0);
+}
+
+bool ThreadStore::ShouldTriggerFinalizingGC()
+{
+    return m_TriggerFinalizingGC;
+}
+
+void ThreadStore::TriggerFinalizingGCIfNecessary()
+{
+    if (!m_TriggerFinalizingGC)
+    {
+        return;
+    }
+    m_TriggerFinalizingGC = false;
+
+    IGCHeap *gcHeap = GCHeapUtilities::GetGCHeap();
+    gcHeap->GarbageCollect(gcHeap->GetMinFinalizingGeneration(), FALSE, collection_non_blocking);
```

Comment:
I thought you wanted this for the idling scenario where nothing is happening aside from the timer creating threads? If there's work going on, you'd have gen1 GCs (or sometimes gen2 GCs) happening anyway so you don't need to worry about this. So if you are triggering GCs, I'd think a gen1 GC would suffice 'cause all the previously created threads would be dead.